fix(core): Ensure backwards-referenced transplanted views are refreshed

atscott opened the pull request:
Based off of https://github.com/angular/angular/pull/51515This commit runs change detection in a loop while there are still dirty
views to be refreshed in the tree. At the moment, this only applies to
transplanted views but will also apply to views with changed signals.

fixes https://github.com/angular/angular/issues/49801

pkozlowski-opensource commented on packages/core/src/render3/interfaces/view.ts:
```diff
@@ -319,12 +324,11 @@ export interface LView<T = unknown> extends Array<any> {
+   * traverse its children during change detection.
    */
-  [DESCENDANT_VIEWS_TO_REFRESH]: number;
+  [HAS_CHILD_VIEWS_TO_REFRESH]: boolean;
```
since it is a boolean now (and _not_ a number) we should move it to flags, reuse one of the bits there and drop this one slot from the LView array

atscott replied:
@pkozlowski-opensource This update is done in the other PR: #51515

Note that the change involves a small bit of extra handling because the `LContainer` also needed this indicator and now the slot isn't shared between `LView` and `LContainer`: https://github.com/angular/angular/pull/51515/commits/2a8e70b2c94c9b11d0d7c015db10bd48cad5778d

AndrewKushnir replied:
nit: we can consider adding a `FLAGS` slot to the `LContainer` as well, since we'll have 2 booleans there (1 existing and 1 newly added in this PR). But this is a refactoring that we can do later in a followup PR.

dylhunn reviewed (approved):
reviewed-for: public-api

jessicajaniuk reviewed (approved):
reviewed-for: public-api

alxhub commented on packages/core/src/render3/instructions/change_detection.ts:
```diff
@@ -37,6 +38,25 @@ export function detectChangesInternal<T>(
+    // run, detect changes on the view again. We run change detection in `Targeted` mode to only
+    // refresh views with the `RefreshView` flag.
+    while (lView[FLAGS] & LViewFlags.RefreshView || lView[HAS_CHILD_VIEWS_TO_REFRESH]) {
+      if (retries === 100) {
```
`100` is a pretty meaningful constant here - let's extract it and give it a name.

alxhub commented on packages/core/src/linker/view_container_ref.ts:
```diff
@@ -506,6 +506,8 @@ const R3ViewContainerRef = class ViewContainerRef extends VE_ViewContainerRef {
 
         prevVCRef.detach(prevVCRef.indexOf(viewRef));
       }
+    } else {
```
Was this just missing before?

atscott replied:
It was one of the things that was missing to make the counters correct for the transplanted views. This was left over from my work to fix that. Reverted.

atscott committed:
fix(core): Ensure backwards-referenced transplanted views are refreshed

This commit runs change detection in a loop while there are still dirty
views to be refreshed in the tree. At the moment, this only applies to
transplanted views but will also apply to views with changed signals.

fixes angular#49801

pkozlowski-opensource commented on packages/core/src/render3/instructions/change_detection.ts:
```diff
@@ -19,6 +20,11 @@ import {getComponentLViewByIndex, isCreationMode, markAncestorsForTraversal, mar
+/**
+ * The maximum number of times the change detection traversal will rerun before throwing an error.
+ */
+const MAXIMUM_REFRESH_RERUNS = 100;
```
As far as I remember we were discussing lowering this number?

atscott replied:
I do not recall this conversation. But the number is an arbitrary decision by me. I have no opinions on what it should be.

atscott commented:
[green TGP](https://fusion2.corp.google.com/presubmit/572708933/OCL:572708933:BASE:572920807:1697126221798:96f4bf8a/targets)

atscott commented:
caretaker: green TGP, current failures are existing failures

dylhunn commented:
This PR was merged into the repository by commit 76152a5fc66e16342045cfd038b53913c32b38da.
